component/encryptor/encryptor.go

```diff
@@ -0,0 +1,382 @@
+package encryptor
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/Azure/azure-storage-fuse/v2/common/config"
+	"github.com/Azure/azure-storage-fuse/v2/common/log"
+	"github.com/Azure/azure-storage-fuse/v2/internal"
+	"github.com/Azure/azure-storage-fuse/v2/internal/handlemap"
+)
+
+type Encryptor struct {
+	internal.BaseComponent
+	handle           *os.File
+	blockSize        uint64
+	mountPointCipher string
+	encryptionKey    []byte
+}
+
+type LastChunkMeta struct {
+	sync.Mutex
+	farthestBlockSeen int64 // keep track of the farthest block that has been written to the file
+	paddingLength     int64 // keep track of the padding length of the last block
+}
+
+var lastchunkMeta = &LastChunkMeta{
+	farthestBlockSeen: -1,
+	paddingLength:     0,
+}
+
+type EncryptorOptions struct {
+	BlockSize          uint64 `config:"block-size-mb" yaml:"block-size-mb,omitempty"`
+	EncryptedMountPath string `config:"encrypted-mount-path" yaml:"encrypted-mount-path,omitempty"`
+	EncryptionKey      string `config:"encryption-key" yaml:"encryption-key,omitempty"`
+}
+
+const (
+	compName    = "encryptor"
+	AuthTagSize = 16
+	NonceSize   = 12
+	MetaSize    = 28
+)
+
+var _ internal.Component = &Encryptor{}
+
+func (e *Encryptor) Name() string {
+	return compName
+}
+
+func (e *Encryptor) SetName(name string) {
+	e.BaseComponent.SetName(name)
+}
+
+func (e *Encryptor) SetNextComponent(nc internal.Component) {
+	e.BaseComponent.SetNextComponent(nc)
+}
+
+func (e *Encryptor) Priority() internal.ComponentPriority {
+	return internal.EComponentPriority.LevelMid()
+}
+
+func (e *Encryptor) Configure(isParent bool) error {
+	log.Trace("Encryptor::Configure :  %s", e.Name())
+
+	// Read the configuration
+	conf := EncryptorOptions{}
+	err := config.UnmarshalKey(e.Name(), &conf)
+	if err != nil {
+		log.Err("Encryptor::Configure : config error [invalid config attributes]")
+		return fmt.Errorf("error reading config for %s: %w", e.Name(), err)
+	}
+
+	// fetch encryption key from environment variable
+	key := os.Getenv("ENCRYPTION_KEY")
+	if key == "" {
+		key = conf.EncryptionKey
+	}
+
+	if key == "" {
+		log.Err("Encryptor::Configure : encryption key not set")
+		return fmt.Errorf("encryption key not set")
+	} else {
+		e.encryptionKey, err = base64.StdEncoding.DecodeString(key)
+		if err != nil {
+			log.Err("Encryptor::Configure : error decoding encryption key")
+			return fmt.Errorf("error decoding encryption key: %w", err)
+		}
+	}
+
+	e.blockSize = 1024 * 1024 // default block size is 1MB
+	if config.IsSet(e.Name() + ".block-size-mb") {
```

Comment:
>if config.IsSet(e.Name() + ".block-size-mb") { [](http://example.com/codeflow?start=1&length=46)

as we have already loaded conf can't the checks be conf.BlockSize !=0 and conf.EncryptedMountPath != "". This is looking odd that even after unmarshalling the configuration we have to peek back into the config to know if value was set or not.

Reply:
yes it did looked odd to me too. But i followed the similar convention blobfuse folks did. 
[https://github.com/abhiguptacse/azure-storage-fuse/blob/0352a942523d179489449bd38a8612e84877f44b/component/block_cache/block_cache.go#L204](https://github.com/abhiguptacse/azure-storage-fuse/blob/0352a942523d179489449bd38a8612e84877f44b/component/block_cache/block_cache.go#L204)

Reply:
Can you move such checks to a separate function say ValidateConfig() instead of having them in configure ? Let configure just work with encryptorOptions.